### ISSUE TITLE
Improve design of GridTools::find_active_cell_around_point()

### ISSUE THREAD

**Issue by nfehn** [dealii/dealii#10535]
This issue discusses design changes of `GridTools::find_active_cell_around_point()`, but not bug fixes:

Let's collect the opinions mentioned so far in other chats:

@nfehn: throwing exceptions

> I realized that most of these tests do not contain a try-block around the function call. The question is, however, whether this behavior should be changed anyway. In my opinion it is not a good design that the function throws if no point is found (because the try-blocks are not used in practice). I would prefer a return type of std::vector, and an empty vector indicates that no point has been found, as do the find_all_...()-type functions introduced later.

@kronbichler: Default parameters

> One thing that is a bit annoying is the fact that the function that already had the tolerance, find_all_active_cells_around_point, uses the arguments in a different order to what is done here. I guess there is little we can do there, but in general we should try to get the interface in an intuitive way. Once we have merged #10477, we should also include that function. That said, I would like to have many opinions on how we want to go ahead with default parameters. 

@tamiko: Backward compatibility

> I just want to add one point to the discussion:

> If possible, we should try to maintain a backward compatible behavior as much as possible. Simply because changing functionality (such as returning a different type, throwing/not throwing an exception) will inevitably break user code.

> As a good metric, we have used our testsuite for precisely that purpose in the past: our tests are a good indicator what semantics the original author had in mind, or at least, what another developer thought is an intended use case. So I feel a bit uneasy about the implied statement that the tests in question are faulty - breaking those might imply that we break a number of (formerly valid) user programs in subtle ways.

@luca-heltai: throwing exceptions

> We should also provide a function that does not throw when points are not found. Suggestions?

**Comment by kronbichler** [comment]
Let me comment on this one:
> I realized that most of these tests do not contain a try-block around the function call. The question is, however, whether this behavior should be changed anyway. In my opinion it is not a good design that the function throws if no point is found (because the try-blocks are not used in practice). I would prefer a return type of std::vector, and an empty vector indicates that no point has been found, as do the find_all_...()-type functions introduced later.

I think the most intuitive C++ way is to return `MeshType::end()` in case we do not find a point, rather than throw an exception. We have seen that things get messy with lots of try/catch blocks. Returning the invalid `end()` iterator would be in line with the typical C++ `find` functions.

I admit that I do not know all internals in every function here and what is expected in user codes, so I can't say I have a solution for how to handle the issue of backward compatibility in case change behavior now. In any case, the considerable number of bug reports and fixes in the last months are indicating that the current state is a bit of a mess. So I guess we should strive to improve the situation and make the functions more robust.

**Comment by bangerth** [comment]
I think the idea of throwing an exception is rooted in the idea that we want to indicate to the caller that we want to return an "exceptional value", not necessarily that "this is an error". But it is correct that we already have an exceptional value, namely the end iterator, and that we should be using that in conformance with `std::find` as @kronbichler correctly points out. 

There is little danger to making such a change because one can not do very much with an end iterator. Furthermore, it is easy to rewrite code in a way that makes it backward compatible: keep the `try-catch` block (no longer needed for the new deal.II, but harmless) *and* deal with end iterators (not needed for older versions, but also harmless).

**Comment by dangars** [comment]
I would like to join the discussion :)

I agree with @tamiko that breaking backwards compatibility is not a good idea. I think that it would be better to use a different name for the new function and deprecate this one.

Step-62 now hangs because of this change. Interestingly throwing an exception when HDF5 is used in parallel hangs the program. A user not familiar with MPI debugging would have a hard time to find out why the program hangs. 

I think that throwing an exception if the point is not found is not a good idea. That is very slow.

**Comment by dangars** [comment]
Have you made a decision regarding `find_active_cell_around_point()` ?

Would it be possible to keep `find_active_cell_around_point()`  and deprecate it in order not to break backwards compatibility? And then add a new function with a different name and the new behavior?

I think that it would be better not to use exceptions when points are not found. It seems that exceptions with parallel HDF5 can freeze deal.II (see #10746 )

Thanks :)

**Comment by luca-heltai** [comment]
We could provide a unified function, with the same name, taking a version of our enumerator and an AdditionalData structure,

~~~C++
enum (class?) PointFinderOptions (name?) {
  throw_if_not_found, 
  ignore_tolerance,
  ignore_artificial,
  ...
} 

struct PointFinderAdditionalData {
   std::vector<bool> selected_vertices;
   double tolerance;
   ...  // etc.
}
~~~
and then use it as
~~~
template<class MeshType>
std::pair<MeshType::iterator, Point<MeshType::space_dimension>> find_active_cell_around_point(const Point<MeshType::space_dimension> &point, const PointFinderOptions & options=reasonable_default, const PointFinderAdditionalData &additional_data=PointFinderAdditionalData(), const Mapping<MeshType::dimension,MeshType::space_dimension> &mapping);
~~~
moving all different implementations in this one single function, making all other functions call this one, and deprecate them.

By default, this one would not throw, would ignore artificial cells, etc. making sure that it behaves as the std::find function.

**Comment by dangars** [comment]
This sounds good to me

**Comment by dangars** [comment]
Would it be possible to revert the function `find_active_cell_around_point()` to the old behavior? After the revert we can decide what to do?

I'm using an old version of deal.II because this change breaks my code. step-62 is also broken.

I'm happy to revert the function to the old behavior. If you agree to revert, I'll do if when you give me the green light :smile:

**Comment by nfehn** [comment]
Let me ask some questions:

I think we agreed that the old behavior has been a bug that can not be tolerated? I am not sure what you mean exactly with revert, or, better, to which version to revert to, or which PR caused problems in your case?

How do you argue that your intended change is an improvement overall (because previous issues, bugfixes, PRs have also been considered an improvement and, finally, have been merged)? The concerns that I have are that reverting to former code will bring up old problems, break other user code, etc.

Since this issue is about improving the design, I would suggest to better use issue #10175 or issues references there. I think we should not always use the most recent issue on find_active_cell_around_point to get the whole problem solved. The discussion here is somewhat stalled also because previous issues on that topic are stalled. My personal point of view is that we have to first fix remaining old problems to get out of this find_active_cell_around_point dilemma, and accept that some user codes have to be adapted simply because the behavior of the function has not been reliable or precisely documented, and has been buggy. In a second step, I would think about design changes, unified interfaces, etc. as suggested by the present issue.

**Comment by dangars** [comment]
The last version of deal.II that works for me is 3d649982db (July 2020) I can not contribute anymore to deal.II because the new behavior of `find_active_cell_around_point()` freezes my code (and also step-62)

Using exceptions for notifications is not a good idea. That has the potential to freeze MPI codes that use MPI_Barriers. These types of bugs are very difficult to find, see #11050 and #10737

I think that it would be better to use a different name for the new function and deprecate `find_active_cell_around_point()` to avoid breaking user codes between deal.II 9.2 and deal.II 9.3

**Comment by dangars** [comment]
The bug is described in #10737

HDF5 can't be used with the new version of `find_active_cell_around_point()`.

It is not just HDF5. Any code that uses MPI Barriers freezes with the new version of `find_active_cell_around_point()`.

**Comment by luca-heltai** [comment]
I'm fixing this.

**Comment by dangars** [comment]
@luca-heltai Thanks :)
